File: 0395_Longest Substring with At Least K Repeating Characters/LongestSubstringwithAtLeastKRepeatingCharacters_slidingwindow.cs
public class Solution {
    public int LongestSubstring(string s, int k) {
        var counters = new int[26];
        var maxUnique = MaxUnique(s);
        int ans = 0;
        
        for(int currUnique = 1; currUnique <= maxUnique; currUnique++)
        {
            Array.Fill(counters, 0);
            var start = 0;
            var end = 0;
            var idx = 0;
            var unique = 0;
            var countAtLeastK = 0;
            
            while(end < s.Length)
            {
                //expand the window
                if(unique <= currUnique)
                {
                    idx = s[end] - 'a';
                    if(counters[idx] == 0) unique++;
                    counters[idx]++;
                    if(counters[idx] == k) countAtLeastK++;
                    end++;
                }
                else
                {
                    //shrink the window
                    idx = s[start] - 'a';
                    if(counters[idx] == k) countAtLeastK--;
                    counters[idx]--;
                    if(counters[idx] == 0) unique--;
                    start++;
                }
                
                if(unique == currUnique && unique == countAtLeastK){
                    ans = Math.Max(ans, end - start);
                }
            }
        }
        
        return ans;
        
    }
    
    private int MaxUnique(string s)
    {
        var count = 0;
        var counters = new bool[26];
        foreach(var c in s)
        {
            if(!counters[c - 'a'])
            {
                count++;
                counters[c - 'a'] = true;
            }
        }
        
        return count;
    }
}

File: 0395_Longest Substring with At Least K Repeating Characters/LongestSubstringwithAtLeastKRepeatingCharacters_slidingwindow.java
class Solution {
    public int longestSubstring(String s, int k) {
        if(s == null || s.length() < k) return 0;
        int[] counters = new int[26];
        int maxUnique =  maxUnique(s);
        int ans = 0;
        
        for(int currUnique = 1; currUnique<=maxUnique; currUnique++) {
            Arrays.fill(counters, 0);
            int start = 0;
            int end = 0;
            int idx = 0;
            int unique = 0;
            int countAtLeastK = 0;
            
            while(end < s.length()) {
                //expend the window
                if(unique <= currUnique)
                {
                    idx = s.charAt(end) - 'a';
                    if(counters[idx] == 0) unique++;
                    counters[idx]++;
                    if(counters[idx] == k) countAtLeastK++;
                    end++;
                }
                else
                {
                    //shrink the window
                    idx = s.charAt(start) - 'a';
                    if(counters[idx] == k) countAtLeastK--;
                    counters[idx]--;
                    if(counters[idx] == 0) unique--;
                    start++;
                }
                
                if(unique == currUnique && unique == countAtLeastK){
                    ans = Math.max(ans, end - start);
                }
            }
        }
        
        return ans;
    }
    
    private int maxUnique(String s) {
        boolean[] check = new boolean[26];
        int ans = 0;
        for(char c : s.toCharArray()) {
            if(!check[c-'a'])
            {
                ans++;
                check[c-'a'] = true;
            }
        }
        
        return ans;
    }
}